modules/gateways/cardinity.php

```diff
@@ -443,16 +377,42 @@ function createWhmcsReturnArray($status, $rawData, $transactionId = '')
  */
 function addRemoteToken($invoiceId, $remoteTokenID)
 {
-    $user = Capsule::table('tblinvoiceitems')->select('userid')
+    if($remoteTokenID){
+        $user = Capsule::table('tblinvoiceitems')->select('userid')
         ->where('invoiceid', $invoiceId)
         ->first();
 
-    Capsule::table('tblclients')
-        ->where('id', $user->userid)
-        ->update([
-            "gatewayid" => $remoteTokenID
-        ]);
+        Capsule::table('tblclients')
+            ->where('id', $user->userid)
+            ->update([
+                "gatewayid" => $remoteTokenID
+            ]);
+    }
 }
 
+/**
+ * Find customers gateway token from invoice
+ *
+ * @param [type] $invoiceId
+ * @return string Cardinity payment ID, used as a token for recurring transactions
+ */
+function getCustomerGatewayId($invoiceId)
+{
+
+    $lastPaidInvoice = Capsule::table('tblinvoiceitems')->select('userid')
+        ->where('invoiceid', $invoiceId)
+        ->first();
+
+    $client = Capsule::table('tblaccounts')
+        ->select('transid')
+        ->where('userid',$lastPaidInvoice->userid)
+        ->where('gateway','cardinity')
+        ->latest('id')
+        ->first();
+
+    return $client->transid;
```

Comment:
its picking the payment method of the card. and capture will call the module. 
if cvv is entered it will use the selected card info (holder name pan date etc...) and make a new payment.

if cvv is empty it will not use the pan number, but use the regular recurring process of creating new payment from rpevious payment_id

Reply:
> if cvv is empty it will not use the pan number, but use the regular recurring process of creating new payment from rpevious payment_id

Will it use last payment_id of the selected card, or last payment_id used in some earlier invoice?

Reply:
last payment id from all of user invoice, it could be a different card too, unfortunately.

i could not find a way to select invoice by the instrument pan as its not stored.